hoot-josm/src/main/cpp/hoot/core/josm/ops/JosmMapValidatorAbstract.cpp

```diff
@@ -0,0 +1,293 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * --------------------------------------------------------------------
+ *
+ * The following copyright notices are generated automatically. If you
+ * have a new notice to add, please use the format:
+ * " * @copyright Copyright ..."
+ * This will properly maintain the copyright information. DigitalGlobe
+ * copyrights will be updated automatically.
+ *
+ * @copyright Copyright (C) 2019, 2020 DigitalGlobe (http://www.digitalglobe.com/)
+ */
+#include "JosmMapValidatorAbstract.h"
+
+// hoot
+#include <hoot/core/util/Log.h>
+#include <hoot/core/jni/JavaEnvironment.h>
+#include <hoot/core/util/StringUtils.h>
+#include <hoot/core/jni/JniConversion.h>
+#include <hoot/core/util/MapProjector.h>
+#include <hoot/core/jni/JniUtils.h>
+
+namespace hoot
+{
+
+JosmMapValidatorAbstract::JosmMapValidatorAbstract() :
+_maxElementsForMapString(2000000),
+_javaEnv(JavaEnvironment::getEnvironment()),
+_josmInterfaceInitialized(false),
+_numValidationErrors(0),
+_numFailingValidators(0)
+{
+  _josmInterfaceName = ConfigOptions().getJosmMapValidatorJavaImplementation();
+}
+
+JosmMapValidatorAbstract::~JosmMapValidatorAbstract()
+{
+  // See related note in _initJosmImplementation.
+  //if (_javaEnv != 0 && _josmInterface != 0)
+  //{
+    // We have a memory leak if this doesn't happen.
+    //LOG_DEBUG("Deleting josm interface...");
+    //_javaEnv->DeleteGlobalRef(_josmInterface);
+  //}
+}
+
+void JosmMapValidatorAbstract::setConfiguration(const Settings& conf)
+{
+  ConfigOptions opts(conf);
+  _josmValidatorsExclude = opts.getJosmValidatorsExclude();
+  _josmValidatorsInclude = opts.getJosmValidatorsInclude();
+  _josmValidatorsRequiringUserCert = opts.getJosmValidatorsRequiringUserCertificate();
+  _maxElementsForMapString = opts.getJosmMaxElementsForMapString();
+}
+
+void JosmMapValidatorAbstract::_initJosmImplementation()
+{
+  LOG_DEBUG("Initializing JOSM implementation...");
+
+  LOG_VART(_josmInterfaceName);
+  jstring interfaceJavaStr = JniConversion::toJavaString(_javaEnv, _josmInterfaceName);
+  //LOG_VART(interfaceJavaStr);
+  jboolean isCopy;
+  const char* interfaceChars = _javaEnv->GetStringUTFChars(interfaceJavaStr, &isCopy);
+  LOG_VART(interfaceChars);
+  _josmInterfaceClass = _javaEnv->FindClass(interfaceChars);
+  LOG_VART(_josmInterfaceClass == 0);
+  _javaEnv->ReleaseStringUTFChars(interfaceJavaStr, interfaceChars);
+
+  _josmInterface =
+    // Thought it would be best to grab this as a global ref, since its possible it could be garbage
+    // collected at any time. Then it can be cleaned up in the destructor. However, that's causing
+    // crashes during the tests, so need to think about it some more and be sure its not needed.
+    /*_javaEnv->NewGlobalRef(*/
```

Comment:
I'm still a little wary of this, but have tested with some large maps and haven't had any memory issues...will keep an eye on it for awhile.